azuredeploy.tf

```diff
@@ -32,78 +45,78 @@ resource "random_string" "name" {
-resource "azurerm_azuread_application" "test" {
-  count                       = "${data.azurerm_client_config.current.service_principal_application_id == "" ? 1 : 0}"
-  name                        = "eventhubstest"
-  homepage                    = "https://eventhubstest"
-  identifier_uris             = ["https://eventhubstest"]
-  reply_urls                  = ["https://eventhubstest"]
-  available_to_other_tenants  = false
-  oauth2_allow_implicit_flow  = true
+resource "azuread_application" "test" {
+  count                      = data.azurerm_client_config.current.service_principal_application_id == "" ? 1 : 0
+  name                       = "eventhubstest"
+  homepage                   = "https://eventhubstest"
+  identifier_uris            = ["https://eventhubstest"]
+  reply_urls                 = ["https://eventhubstest"]
+  available_to_other_tenants = false
+  oauth2_allow_implicit_flow = true
 }
 
 # Create a service principal, which represents a linkage between the AAD application and the password
-resource "azurerm_azuread_service_principal" "test" {
-  count          = "${data.azurerm_client_config.current.service_principal_application_id == "" ? 1 : 0}"
-  application_id = "${azurerm_azuread_application.test.application_id}"
+resource "azuread_service_principal" "test" {
+  count          = data.azurerm_client_config.current.service_principal_application_id == "" ? 1 : 0
+  application_id = azuread_application.test[0].application_id
 }
 
 # Create a new service principal password which will be the AZURE_CLIENT_SECRET env var
-resource "azurerm_azuread_service_principal_password" "test" {
-  count                 = "${data.azurerm_client_config.current.service_principal_application_id == "" ? 1 : 0}"
-  service_principal_id  = "${azurerm_azuread_service_principal.test.id}"
-  value                 = "${random_string.secret.result}"
-  end_date              = "2030-01-01T01:02:03Z"
+resource "azuread_service_principal_password" "test" {
+  count                = data.azurerm_client_config.current.service_principal_application_id == "" ? 1 : 0
+  service_principal_id = azuread_service_principal.test[0].id
+  value                = random_string.secret[0].result
+  end_date             = "2030-01-01T01:02:03Z"
 }
 
 # This provides the new AAD application the rights to managed, send and receive from the Event Hubs instance
 resource "azurerm_role_assignment" "service_principal_eh" {
-  count                 = "${data.azurerm_client_config.current.service_principal_application_id == "" ? 1 : 0}"
-  scope                 = "subscriptions/${data.azurerm_client_config.current.subscription_id}/resourceGroups/${azurerm_resource_group.test.name}/providers/Microsoft.EventHub/namespaces/${azurerm_eventhub_namespace.test.name}"
-  role_definition_name  = "Owner"
-  principal_id          = "${azurerm_azuread_service_principal.test.id}"
+  count                = data.azurerm_client_config.current.service_principal_application_id == "" ? 1 : 0
```

Comment:
I think this should be run no matter what. Since, the service principal will always need DataOwner access whether it's provided to the template or if it's created in the template (count 0 or 1).

I recommend removing line 104.